### PR TITLE
Added tests for send and receive

### PR DESCRIPTION
**Pull request by SRYER**
I have added initial tests prior to changing/adding any new features. I added a test framework that allows tests for being run with "cargo nextest run" and show test result in a more concise manner compared to "cargo test".

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -15,3 +15,8 @@ tokio-tungstenite = "0.18"
 tokio = { version = "1.3", features = ["full"] }
 futures-util = "0.3"
 flume = "0.10"
+
+[dev-dependencies]
+rstest = "0.16.0"
```
Comment (Flying-Toast) [inline Cargo.toml:20]: What do we need rstest for? As far as I can see, it's not referenced other than here.

File: tests/tests.rs [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    use simple_websockets;
+    use url::Url;
+
+    #[test]
+    fn connect_disconnect_test() {
+        // Start a server
+        const UNIQUE_TEST_PORT: u16 = 9000; // Future fix will eliminate this need
+        let server_endpoint = Url::parse(format!("ws://127.0.0.1:{UNIQUE_TEST_PORT}").as_str()).unwrap();
+        let websocket_event_hub = simple_websockets::launch(UNIQUE_TEST_PORT).expect(format!("failed to listen on websocket port {UNIQUE_TEST_PORT}").as_str());
+        assert_eq!(true, websocket_event_hub.is_empty());
```
Comment (Flying-Toast) [inline tests/tests.rs:13]: `assert_eq!(true, ...)` -> `assert!(...)`

File: tests/tests.rs [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+        let (mut client_0, _response_0) = tungstenite::connect(&server_endpoint).expect("Can't connect");
+        std::thread::sleep(std::time::Duration::from_millis(500)); // Ensure event is actually triggered. The longer the wait, the slower test. The short the wait the higher risk of getting an unstable test. Optimal solution would be for 
+        assert_connect_event(websocket_event_hub.poll_event(), 0);
+        assert_eq!(true, websocket_event_hub.is_empty());
+
+        let (mut client_1, _response_1) = tungstenite::connect(&server_endpoint).expect("Can't connect");
+        let (mut _client_2, _response_2) = tungstenite::connect(&server_endpoint).expect("Can't connect");
+
+        std::thread::sleep(std::time::Duration::from_millis(500));
+        assert_eq!(false, websocket_event_hub.is_empty());
```
Comment (Flying-Toast) [inline tests/tests.rs:25]: `assert_eq!(false, ...)` -> `assert!(!...)`

File: Cargo.toml (hunk 1 above)
Comment (SRYER) [reply Cargo.toml:20]: That's what make "cargo nextest run" possible. Output from this command:

cargo nextest run
   Compiling simple-websockets v0.1.4 (C:\git_thirdparty\simple-websockets)
    Finished test [unoptimized + debuginfo] target(s) in 0.78s
    Starting 5 tests across 2 binaries
        PASS [   1.030s] simple-websockets::tests tests::receive_text_message_test
        PASS [   1.026s] simple-websockets::tests tests::send_text_message_test
        PASS [   1.033s] simple-websockets::tests tests::receive_binary_message_test
        PASS [   1.029s] simple-websockets::tests tests::send_binary_message_test
        PASS [   2.552s] simple-websockets::tests tests::connect_disconnect_test
------------
     Summary [   2.553s] 5 tests run: 5 passed, 0 skipped
     
Output from standard "cargo test":

cargo test       
    Finished test [unoptimized + debuginfo] target(s) in 0.11s
     Running unittests src\lib.rs (target\debug\deps\simple_websockets-c84e5e08244f78c3.exe) 

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests\tests.rs (target\debug\deps\tests-3ea366ce12077e54.exe)

running 5 tests
test tests::send_text_message_test ... ok     
test tests::receive_text_message_test ... ok  
test tests::send_binary_message_test ... ok   
test tests::receive_binary_message_test ... ok
test tests::connect_disconnect_test ... ok

test result: ok. 5 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 2.54s

   Doc-tests simple-websockets

running 1 test
test src\lib.rs - (line 10) - compile ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.06s


If you don't want this one, I can remove it and just add it while I work and remove it before I commit. But it really makes for a better test overview imo :)

**Comment by Flying-Toast** [comment]
I left a few comments. When I run the tests, all 5 are failing:

```
---- tests::connect_disconnect_test stdout ----
thread 'tests::connect_disconnect_test' panicked at 'Can't connect: Url(UnableToConnect("ws://127.0.0.1:9000/"))', tests/tests.rs:16:82

---- tests::send_binary_message_test stdout ----
thread 'tests::send_binary_message_test' panicked at 'Can't connect: Url(UnableToConnect("ws://127.0.0.1:9004/"))', tests/tests.rs:146:65

---- tests::receive_binary_message_test stdout ----
thread 'tests::receive_binary_message_test' panicked at 'Can't connect: Url(UnableToConnect("ws://127.0.0.1:9002/"))', tests/tests.rs:84:69
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- tests::receive_text_message_test stdout ----
thread 'tests::receive_text_message_test' panicked at 'Can't connect: Url(UnableToConnect("ws://127.0.0.1:9001/"))', tests/tests.rs:55:69

---- tests::send_text_message_test stdout ----
thread 'tests::send_text_message_test' panicked at 'Can't connect: Url(UnableToConnect("ws://127.0.0.1:9003/"))', tests/tests.rs:114:65
```

File: Cargo.toml (hunk 1 above)
Comment (Flying-Toast) [reply Cargo.toml:20]: Ah, I see. I'd prefer if we just left it with the default `cargo test`, in order to keep dependencies minimal :)

**Comment by SRYER** [comment]
Ohhhh, what environment do you run it in? As it is the connect step, I guess the bind went fine, so my guess it is a local firewall not allowing for connecting to localhost? Maybe using "localhost" may work over "127.0.0.1" - could you give that a try?

**Comment by Flying-Toast** [comment]
Hmm, doesn't seem to be the problem. localhost doesn't work in the test either. The server *does* get started though, I can confirm I can `netcat localhost {UNIQUE_TEST_PORT}`.

**Comment by SRYER** [comment]
What OS/Build environment do you execute the cargo test command in?
As a narrow-down process, does it work if you disable your firewall?

**Comment by Flying-Toast** [comment]
Ah, the issue is that the tests are trying to connect before the thread running the server has had time to start up. For now, adding a short delay after all the calls to `launch()` in the tests will fix it - things seem to be working for me with a `thread::sleep(500ms)` after launch().

**Comment by SRYER** [comment]
Ohhh, great you found the reason! Even though sleeping is never 100% stable, as one day the bind may take longer than that, for 99.9% of the time it will be sufficient, I guess. And with a low number of tests, I hope the total run time is acceptable?

Should I add the sleep to the PR?

**Comment by Flying-Toast** [comment]
Yeah, lets add the sleep to the PR. I've been planning some rework on simple-websockets' API for a while that would remove a lot of these rough edges, but until then let's just leave the delay :)

**Comment by Flying-Toast** [comment]
Looks good now. Thanks :)

**Pull request merged**